pkg/kubelet/kubelet.go

```diff
@@ -1809,15 +1810,11 @@ func (kl *Kubelet) syncPod(_ context.Context, updateType kubetypes.SyncPodType,
 		return false, err
 	}
 
-	// Volume manager will not mount volumes for terminating pods
-	// TODO: once context cancellation is added this check can be removed
-	if !kl.podWorkers.IsPodTerminationRequested(pod.UID) {
-		// Wait for volumes to attach/mount
-		if err := kl.volumeManager.WaitForAttachAndMount(pod); err != nil {
-			kl.recorder.Eventf(pod, v1.EventTypeWarning, events.FailedMountVolume, "Unable to attach or mount volumes: %v", err)
-			klog.ErrorS(err, "Unable to attach or mount volumes for pod; skipping pod", "pod", klog.KObj(pod))
-			return false, err
-		}
+	// Wait for volumes to attach/mount
+	if err := kl.volumeManager.WaitForAttachAndMount(ctx, pod); err != nil {
```

Comment:
The new `wait` methods that returned context.Canceled were intended to avoid writing an error here - if the context is canceled, it's not an "eventable" / loggable error, it's just "we're going to delete".  I was concerned about that excessive logging and thus was trying to avoid it.  The event would be sent whenever pods with volumes got deleted which would be not infrequent during deployment / node upgrades.  I'm open to arguing about whether a few extra errors exposed to the user are better or worse than the delay, but I do think we need to avoid excessive error reporting in that case.  So my preference is still to get the new wait methods in and switch to using them here before we cancel the context, but if @liggitt gets to it soon I think there's a chance to get it in 1.27.

Reply:
SGTM, happy to rework once those go in. Just added a e2e test for this particular scenario.

Reply:
Trying to get the prereq PR in today - if it doesn't, i'll review this as is and we can use this to gather data on the larger context change (since it is reasonably scoped and a prereq).

Reply:
Ok, prereq landed.  I would eat a context canceled error here and return false immediately, and below try to avoid logging anything in a cancel context.

Reply:
I started looking at this, but I'm not entirely sure. I might be missing something, but I think we want to know the difference between a) the 2 minute mount timeout failed and b) the context was cancelled.  Using `wait.PollUntilContextTimeout` makes those indistinguishable from what I can tell. In case a, we should report an error and in b it's fine to report no errors.